Add "channels:archived" search filter.

whilstsomebody opened the pull request:
This PR adds "channels:archived" search filter to make search in archived channels easy.

## Screenshots

<img width="1074" height="380" alt="image" src="https://github.com/user-attachments/assets/1db3fcb2-534c-4474-8320-5f778b64a70e" />

![image](https://github.com/user-attachments/assets/530f5a9a-4231-4a81-8b2f-eac89362dd1b)

![image](https://github.com/user-attachments/assets/1a1e42a6-fd3c-4f37-b2d4-c09a61452cc8)

<img width="814" height="463" alt="image" src="https://github.com/user-attachments/assets/80b92e7d-53c4-4b2c-a0d8-5b5702ab0fce" />


Fixes: #32506

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

apoorvapendse commented on web/src/search_suggestion.ts:
```diff
@@ -184,6 +187,29 @@ function get_channel_suggestions(last: NarrowTerm, terms: NarrowTerm[]): Suggest
+    last: NarrowTerm,
+    terms: NarrowTerm[],
+): Suggestion[] {
+    const archived_description_html = "Archived channels";
```
Should we have `All archived channels` here for consistency with `channels:public` and given that we have this description.
![image](https://github.com/user-attachments/assets/c4f9b7c0-934c-486f-bab8-4cc9f5580736)

apoorvapendse reviewed (commented):
Works in manual testing and changes look good to me.
Posted one review comment.

apoorvapendse commented:
@zulipbot remove "buddy review"

sanchi-t commented on web/src/search_suggestion.ts:
```diff
@@ -142,7 +142,10 @@ function compare_by_direct_message_group(
 }
 
-function get_channel_suggestions(last: NarrowTerm, terms: NarrowTerm[]): Suggestion[] {
+function get_public_channel_filter_suggestions(
+    last: NarrowTerm,
+    terms: NarrowTerm[],
+): Suggestion[] {
```
I think this function is also responsible for fetching suggestions for private channels. Could you check that? If so, the function name might not be accurate.

sanchi-t replied:
I think this function also handles filter suggestions when searching for a single archived channel name. Should we consider renaming the function to better reflect that?

sanchi-t commented on web/templates/search_operators.hbs:
```diff
@@ -70,6 +70,16 @@
                             {{/if}}
                         </td>
                     </tr>
+                    <tr>Add commentMore actions
```
What is this for?

sanchi-t commented on zerver/lib/streams.py:
```diff
@@ -870,6 +870,12 @@ def get_public_streams_queryset(realm: Realm) -> QuerySet[Stream]:
     return Stream.objects.filter(realm=realm, invite_only=False, history_public_to_subscribers=True)
 
 
+def get_archived_streams_queryset(realm: Realm) -> QuerySet[Stream]:
+    return Stream.objects.filter(
+        deactivated=True,
+    )
+
+
```
Shouldn't we fetch archived streams only for the user's realm?

sanchi-t reviewed (approved):
LGTM! Ready for further round of review.

alya commented:
@evykassirer Are you up for reviewing this one?

evykassirer commented on web/src/search_suggestion.ts:
```diff
@@ -184,6 +184,29 @@ function get_channel_suggestions(last: NarrowTerm, terms: NarrowTerm[]): Suggest
     });
 }
 
+function get_archived_channel_filter_suggestions(
```
I would either include this with `get_channels_filter_suggestions` or rename that to `get_public_channels_filter_suggestions`

evykassirer replied:
Oh seeing [this conversation](https://github.com/zulip/zulip/pull/34799#discussion_r2127846327) now. I think ideally you could get them all into `get_channels_filter_suggestions`, but I'm not sure how difficult that would be.

evykassirer commented on web/tests/filter.test.cjs:
```diff
@@ -1671,6 +1726,14 @@ test("describe", ({mock_template, override}) => {
     page_params.is_spectator = false;
 
+    narrow = [{operator: "channels", operand: "archived"}];
+    string = "channels archived";
```
I think "archived channels" would sound better here -- what do you think?

evykassirer replied:
yes that looks great, thanks!

evykassirer reviewed (approved):
Looks good to me, thanks for working on this!

timabbott commented on web/src/filter.ts:
```diff
@@ -878,6 +885,14 @@ export class Filter {
+        if (page_params.is_spectator || current_user.is_guest) {
+            return possible_prefix + "all archived channels that you can view";
+        }
+        return possible_prefix + "all archived channels";
```
I think this one might just want to be "archived channels", without the "all" or conditional. Unlike "public channels", normal users can't access random private channels.

timabbott replied:
Yeah, let's write this as just a fixed "archived channels" string, that's the correct description for this operator.

timabbott commented on web/src/filter.ts:
```diff
@@ -1594,7 +1618,11 @@ export class Filter {
+        if (
+            this.has_operator("channels") ||
+            this.has_negated_operand("channels", "public") ||
+            this.has_negated_operand("channels", "archived")
```
Hmm, I thought someone was fixing this TODO...

timabbott commented on web/src/search_suggestion.ts:
```diff
@@ -668,6 +668,37 @@ function get_public_channels_filter_suggestions(
+                {operator: "dm-including"},
+                {operator: "dm"},
+                {operator: "in"},
+                {operator: "channels"},
```
This isn't incompatible, you can have `channels:public` `channels:archived`. I don't think `in:` is incompatible either.

timabbott commented on web/templates/search_operators.hbs:
```diff
@@ -70,6 +70,16 @@
+                            {{#if can_access_all_public_channels }}
+                            {{t 'Search all archived channels.'}}
+                            {{else}}
+                            {{t 'Search all archived channels that you can view.'}}
```
Again this should probably be just "Search archived channels".

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -1164,6 +1236,18 @@ def test_ok_to_include_history(self) -> None:
         ]
         self.assertFalse(ok_to_include_history(narrow, user_profile, False))
 
+        # channels:archived searches should not include history for non-guest members.Add commentMore actions
```
".Add commentMore actions" seems like a mistake on this line.

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -1237,6 +1321,30 @@ def test_ok_to_include_history(self) -> None:
+            NarrowParameter(operator="channels", operand="archived"),
+            NarrowParameter(operator="is", operand="resolved"),
+        ]
+        self.assertFalse(ok_to_include_history(narrow, user_profile, False))
```
I think this should probably test that if you had `channels:public` too, `ok_to_include_history` becomes true.

(The question is mainly about whether any channels with protected history can exist).

whilstsomebody commented:
@timabbott , I have updated the PR as per your reviews above. Following is the summary of the current state of the PR:

- Updated the screenshots to reflect the new changes.
- Updated all the places mentioning "all archived channels" to just mention "archived channels" including phrase and statements: [1](https://github.com/zulip/zulip/pull/34799/files#diff-80e9489ac91f3b598737b538b4fda1b42ee8f14992937ef4c2afc72106b54ce4R91-R92);,[2](https://github.com/zulip/zulip/pull/34799/files#diff-8bfe95719769a8bcfccf82f67ac69fdaa502d2763889391eac75212297113d6cR893-R897), [3](https://github.com/zulip/zulip/pull/34799/files#diff-8bfe95719769a8bcfccf82f67ac69fdaa502d2763889391eac75212297113d6cR1516-R1517);,[4](https://github.com/zulip/zulip/pull/34799/files#diff-6f0676a8dfe28aab706f372dcaafe6d28cdd59d199d2c843eb937d34a6cb8c5aR73-R78); including tests as well.
- https://github.com/zulip/zulip/pull/34799#discussion_r2224300890: Added the test [here](https://github.com/zulip/zulip/pull/34799/files#diff-fc80cf5301dfd792586bacc6621a3dc0c2aa1fecfb0f7bfd4a5bd8fa5dbd0eb0R1348-R1352).

Could you take a look again? Thanks!

alya commented:
@opmkumar This seems like a good one for you to take over an push to completion.
